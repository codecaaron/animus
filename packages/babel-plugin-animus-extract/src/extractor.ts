import type { NodePath, types as BabelTypes } from '@babel/core';
import type { CallExpression } from '@babel/types';

import { generateCSS } from './css-generator';
import { createRuntimeReplacement } from './transformer';
import type { AnimusExtractState, ComponentData, ExtractedChainCall } from './types';
import { generateComponentId } from './utils';

export function isAnimusBuilder(
  path: NodePath<CallExpression>,
  state: AnimusExtractState
): boolean {
  const { types: t } = state;
  const {callee} = path.node;

  // Direct call: animus.styles()
  if (
    t.isMemberExpression(callee) &&
    t.isIdentifier(callee.object) &&
    (callee.object.name === 'animus' || callee.object.name === state.animusBinding)
  ) {
    return true;
  }

  // Chained call: animus.styles().states()
  if (t.isMemberExpression(callee) && t.isCallExpression(callee.object)) {
    const objectPath = path.get('callee.object') as NodePath<CallExpression>;
    return isAnimusBuilder(objectPath, state);
  }

  return false;
}

function isAnimusChainCall(path: NodePath<CallExpression>, state: AnimusExtractState): boolean {
  const { types: t } = state;
  const {callee} = path.node;

  return t.isMemberExpression(callee) &&
    (isAnimusBuilder(path, state) ||
     (t.isCallExpression(callee.object) && isAnimusChainCall(path.get('callee.object') as NodePath<CallExpression>, state)));
}

export function extractAnimusComponent(
  path: NodePath<CallExpression>,
  state: AnimusExtractState
): void {
  const chain = extractChain(path, state);
  if (chain.length === 0) return;

  const componentData = processChain(chain, state);

  // Generate CSS for this component
  const css = generateCSS(componentData, state);
  if (css) {
    state.cssOutput.set(componentData.id, css);
  }

  // Store component data for JSX transformation
  if (componentData.name) {
    state.animusComponents.set(componentData.name, componentData);
  }

  // Replace the original expression with runtime version
  const replacement = createRuntimeReplacement(componentData, state);
  path.replaceWith(replacement);
}

function extractChain(
  path: NodePath<CallExpression>,
  state: AnimusExtractState
): ExtractedChainCall[] {
  const chain: ExtractedChainCall[] = [];
  let current: NodePath<CallExpression> | null = path;

  while (current && isAnimusChainCall(current, state)) {
    const { types: t } = state;
    const {callee} = current.node;

    if (t.isMemberExpression(callee) && t.isIdentifier(callee.property)) {
      const methodName = callee.property.name;
      const args = current.node.arguments;

      chain.unshift({ methodName, args });

      if (t.isCallExpression(callee.object)) {
        current = current.get('callee.object') as NodePath<CallExpression>;
      } else {
        current = null;
      }
    } else {
      current = null;
    }
  }

  return chain;
}

function processChain(
  chain: ExtractedChainCall[],
  state: AnimusExtractState
): ComponentData {
  const { types: t } = state;
  const componentData: ComponentData = {
    id: generateComponentId(),
  };

  for (const { methodName, args } of chain) {
    switch (methodName) {
      case 'styles':
        if (args[0] && t.isObjectExpression(args[0])) {
          componentData.styles = extractObjectLiteral(args[0], state);
        }
        break;

      case 'states':
        if (args[0] && t.isObjectExpression(args[0])) {
          componentData.states = extractObjectLiteral(args[0], state);
        }
        break;

      case 'variants':
        if (args[0] && t.isObjectExpression(args[0])) {
          componentData.variants = extractObjectLiteral(args[0], state);
        }
        break;

      case 'groups':
        if (args[0] && t.isObjectExpression(args[0])) {
          componentData.groups = extractObjectLiteral(args[0], state);
        }
        break;

      case 'props':
        if (args[0] && t.isObjectExpression(args[0])) {
          componentData.props = extractObjectLiteral(args[0], state);
        }
        break;

      case 'asElement':
        if (args[0] && t.isStringLiteral(args[0])) {
          componentData.element = args[0].value;
        }
        break;

      case 'displayName':
        if (args[0] && t.isStringLiteral(args[0])) {
          componentData.name = args[0].value;
        }
        break;
    }
  }

  return componentData;
}

function extractObjectLiteral(
  node: BabelTypes.ObjectExpression,
  state: AnimusExtractState
): Record<string, any> {
  const { types: t } = state;
  const result: Record<string, any> = {};

  for (const prop of node.properties) {
    if (t.isObjectProperty(prop) || t.isObjectMethod(prop)) {
      const key = getPropertyKey(prop, state);
      if (key && t.isObjectProperty(prop)) {
        const value = extractValue(prop.value, state);
        if (value !== undefined) {
          result[key] = value;
        }
      }
    }
  }

  return result;
}

function getPropertyKey(
  prop: BabelTypes.ObjectProperty | BabelTypes.ObjectMethod,
  state: AnimusExtractState
): string | null {
  const { types: t } = state;

  if (t.isIdentifier(prop.key)) {
    return prop.key.name;
  } else if (t.isStringLiteral(prop.key)) {
    return prop.key.value;
  }

  return null;
}

function extractValue(node: BabelTypes.Node, state: AnimusExtractState): any {
  const { types: t } = state;

  if (t.isStringLiteral(node)) {
    return node.value;
  } else if (t.isNumericLiteral(node)) {
    return node.value;
  } else if (t.isBooleanLiteral(node)) {
    return node.value;
  } else if (t.isObjectExpression(node)) {
    return extractObjectLiteral(node, state);
  } else if (t.isArrayExpression(node)) {
    return node.elements.map(el => el ? extractValue(el, state) : null);
  } else if (t.isUnaryExpression(node) && node.operator === '-' && t.isNumericLiteral(node.argument)) {
    return -node.argument.value;
  }

  // For dynamic values, return undefined to skip extraction
  return undefined;
}
