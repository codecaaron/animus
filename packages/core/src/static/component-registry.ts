import type ts from 'typescript';

import {
  ComponentIdentity,
  ExtractedStylesWithIdentity,
} from './component-identity';
import { CrossFileUsageCollector, GlobalUsageMap } from './cross-file-usage';
import { ImportResolver } from './import-resolver';
import type { UsageMap } from './types';
import { TypeScriptExtractor } from './typescript-extractor';

/**
 * Component entry in the registry
 */
export interface ComponentEntry {
  identity: ComponentIdentity;
  styles: ExtractedStylesWithIdentity;
  lastModified: number;
  dependencies: ComponentIdentity[]; // Components this one extends or references
  dependents: Set<string>; // File paths that use this component
}

/**
 * Registry events for external listeners
 */
export interface RegistryEvents {
  componentAdded: (entry: ComponentEntry) => void;
  componentUpdated: (entry: ComponentEntry) => void;
  componentRemoved: (identity: ComponentIdentity) => void;
  fileInvalidated: (filePath: string) => void;
}

/**
 * The Component Registry - Central authority for all components in the project
 * It sees all, knows all, and maintains the truth across the void
 */
export class ComponentRegistry {
  private components = new Map<string, ComponentEntry>(); // hash -> entry
  private fileComponents = new Map<string, Set<string>>(); // filePath -> Set<hash>
  private extractor: TypeScriptExtractor;
  private resolver: ImportResolver;
  private usageCollector: CrossFileUsageCollector;
  private listeners: Partial<RegistryEvents> = {};

  constructor(private program: ts.Program) {
    this.extractor = new TypeScriptExtractor();
    this.extractor.initializeProgram(this.getProjectRoot());
    this.resolver = new ImportResolver(program);
    this.usageCollector = new CrossFileUsageCollector(program, this.resolver);
  }

  /**
   * Initialize registry by scanning all project files
   */
  async initialize(): Promise<void> {
    const componentFiles = this.extractor.getComponentFiles();

    for (const file of componentFiles) {
      await this.processFile(file);
    }

    // Build dependency relationships
    this.buildDependencyGraph();
  }

  /**
   * Process a single file and register its components
   */
  async processFile(filePath: string): Promise<void> {
    const components = this.extractor.extractFromFile(filePath);
    const fileHashes = new Set<string>();

    for (const component of components) {
      const entry: ComponentEntry = {
        identity: component.identity,
        styles: component,
        lastModified: Date.now(),
        dependencies: [],
        dependents: new Set(),
      };

      // Handle extends relationship
      if (component.extends) {
        entry.dependencies.push(component.extends);
      }

      // Register the component
      this.registerComponent(entry);
      fileHashes.add(component.identity.hash);
    }

    // Track which components are in this file
    this.fileComponents.set(filePath, fileHashes);
  }

  /**
   * Register a component in the registry
   */
  private registerComponent(entry: ComponentEntry): void {
    const existing = this.components.get(entry.identity.hash);

    if (existing) {
      // Update existing
      this.components.set(entry.identity.hash, entry);
      this.emit('componentUpdated', entry);
    } else {
      // Add new
      this.components.set(entry.identity.hash, entry);
      this.emit('componentAdded', entry);
    }
  }

  /**
   * Get a component by its identity
   */
  getComponent(identity: ComponentIdentity): ComponentEntry | undefined {
    return this.components.get(identity.hash);
  }

  /**
   * Get all components
   */
  getAllComponents(): ComponentEntry[] {
    return Array.from(this.components.values());
  }

  /**
   * Get components from a specific file
   */
  getFileComponents(filePath: string): ComponentEntry[] {
    const hashes = this.fileComponents.get(filePath);
    if (!hashes) return [];

    return Array.from(hashes)
      .map((hash) => this.components.get(hash))
      .filter((c): c is ComponentEntry => c !== undefined);
  }

  /**
   * Get usage map for a component
   */
  getComponentUsage(identity: ComponentIdentity): UsageMap {
    return this.usageCollector.buildComponentUsageMap(identity);
  }

  /**
   * Get global usage map for all components
   */
  getGlobalUsage(): GlobalUsageMap {
    return this.usageCollector.collectFromProgram();
  }

  /**
   * Invalidate a file and reprocess it
   */
  async invalidateFile(filePath: string): Promise<void> {
    // Remove old components from this file
    const oldHashes = this.fileComponents.get(filePath);
    if (oldHashes) {
      for (const hash of oldHashes) {
        const component = this.components.get(hash);
        if (component) {
          this.components.delete(hash);
          this.emit('componentRemoved', component.identity);
        }
      }
    }

    // Clear usage cache
    this.usageCollector.invalidateFile(filePath);

    // Reprocess the file
    await this.processFile(filePath);

    // Rebuild dependencies for affected components
    this.updateDependentsOf(filePath);

    this.emit('fileInvalidated', filePath);
  }

  /**
   * Build the dependency graph between components
   */
  private buildDependencyGraph(): void {
    // Clear existing dependents
    for (const entry of this.components.values()) {
      entry.dependents.clear();
    }

    // Build dependents from dependencies
    for (const entry of this.components.values()) {
      for (const dep of entry.dependencies) {
        const depEntry = this.getComponent(dep);
        if (depEntry) {
          depEntry.dependents.add(entry.identity.filePath);
        }
      }
    }

    // Add usage-based dependents
    const globalUsage = this.getGlobalUsage();
    for (const [hash, usage] of globalUsage) {
      const component = this.components.get(hash);
      if (component) {
        for (const use of usage.usages) {
          component.dependents.add(use.usageLocation);
        }
      }
    }
  }

  /**
   * Update components that depend on a changed file
   */
  private updateDependentsOf(filePath: string): void {
    const changedComponents = this.getFileComponents(filePath);

    for (const component of changedComponents) {
      for (const dependentFile of component.dependents) {
        // Mark dependent files as needing reprocessing
        // In a real implementation, this would trigger incremental compilation
        this.usageCollector.invalidateFile(dependentFile);
      }
    }
  }

  /**
   * Get components sorted by extension hierarchy (topological sort)
   * Parents come before children to ensure proper CSS cascade
   */
  getComponentsSortedByExtension(): ComponentEntry[] {
    const components = Array.from(this.components.values());
    const visited = new Set<string>();
    const visiting = new Set<string>();
    const sorted: ComponentEntry[] = [];

    const visit = (component: ComponentEntry): void => {
      const hash = component.identity.hash;

      if (visiting.has(hash)) {
        // Circular dependency detected - skip to prevent infinite loop
        return;
      }

      if (visited.has(hash)) {
        return;
      }

      visiting.add(hash);

      // Visit dependencies first (parents before children)
      for (const dep of component.dependencies) {
        const depComponent = this.getComponent(dep);
        if (depComponent) {
          visit(depComponent);
        }
      }

      visiting.delete(hash);
      visited.add(hash);
      sorted.push(component);
    };

    // Visit all components
    for (const component of components) {
      if (!visited.has(component.identity.hash)) {
        visit(component);
      }
    }

    return sorted;
  }

  /**
   * Get the project root directory
   */
  private getProjectRoot(): string {
    const files = this.program.getRootFileNames();
    if (files.length === 0 || files.some((file) => file.includes('vite-test'))) return process.cwd();

    // Find common ancestor directory

    const path = require('path');
    let commonDir = path.dirname(files[0]);

    for (const file of files) {
      const dir = path.dirname(file);
      while (!dir.startsWith(commonDir)) {
        commonDir = path.dirname(commonDir);
      }
    }

    return commonDir;
  }

  /**
   * Subscribe to registry events
   */
  on<K extends keyof RegistryEvents>(
    event: K,
    handler: RegistryEvents[K]
  ): void {
    this.listeners[event] = handler;
  }

  /**
   * Emit an event
   */
  private emit<K extends keyof RegistryEvents>(
    event: K,
    ...args: Parameters<RegistryEvents[K]>
  ): void {
    const handler = this.listeners[event];
    if (handler) {
      (handler as any)(...args);
    }
  }

  /**
   * Get registry statistics
   */
  getStats(): {
    totalComponents: number;
    totalFiles: number;
    componentsWithUsage: number;
    componentsWithDependents: number;
    averageDependentsPerComponent: number;
  } {
    const stats = {
      totalComponents: this.components.size,
      totalFiles: this.fileComponents.size,
      componentsWithUsage: 0,
      componentsWithDependents: 0,
      averageDependentsPerComponent: 0,
    };

    let totalDependents = 0;

    for (const component of this.components.values()) {
      if (component.dependents.size > 0) {
        stats.componentsWithDependents++;
        totalDependents += component.dependents.size;
      }
    }

    // Check usage
    const globalUsage = this.getGlobalUsage();
    stats.componentsWithUsage = globalUsage.size;

    if (stats.componentsWithDependents > 0) {
      stats.averageDependentsPerComponent =
        totalDependents / stats.componentsWithDependents;
    }

    return stats;
  }

  /**
   * Clear all caches and reset the registry
   */
  clear(): void {
    this.components.clear();
    this.fileComponents.clear();
    this.usageCollector.clearCache();
    this.resolver.clearCache();
  }
}

// The Registry is manifest - the central authority across the void
// All components are known, their relationships tracked, their usage recorded
