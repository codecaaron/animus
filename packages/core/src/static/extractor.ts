import * as parser from '@babel/parser';
// Handle the babel/traverse CommonJS export issue
// @ts-ignore - babel/traverse has complex module exports
import traverseDefault from '@babel/traverse';
import * as t from '@babel/types';

const traverse = (traverseDefault as any).default || traverseDefault;

import type { NodePath } from '@babel/traverse';

/**
 * Extracted style data from an Animus component
 */
export interface ExtractedStyles {
  componentName?: string;
  baseStyles?: Record<string, any>;
  variants?: Record<string, any> | Record<string, any>[];
  states?: Record<string, any>;
  groups?: string[];
  props?: Record<string, any>;
}

/**
 * Extract styles from Animus method chains in code
 */
export function extractStylesFromCode(code: string): ExtractedStyles[] {
  const ast = parser.parse(code, {
    sourceType: 'module',
    plugins: ['typescript', 'jsx'],
  });

  const extractedComponents: ExtractedStyles[] = [];

  traverse(ast as any, {
    CallExpression(path: NodePath<t.CallExpression>) {
      // Look for .styles() calls
      if (
        t.isMemberExpression(path.node.callee) &&
        t.isIdentifier(path.node.callee.property, { name: 'styles' })
      ) {
        const extracted: ExtractedStyles = {};

        // Try to get component name from variable declaration
        const varDeclarator = path.findParent((p: NodePath) =>
          p.isVariableDeclarator()
        );
        if (varDeclarator && t.isVariableDeclarator(varDeclarator.node)) {
          if (t.isIdentifier(varDeclarator.node.id)) {
            extracted.componentName = varDeclarator.node.id.name;
          }
        }

        // Extract styles object from first argument
        const stylesArg = path.node.arguments[0];
        if (t.isObjectExpression(stylesArg)) {
          extracted.baseStyles = extractObjectLiteral(stylesArg);
        }

        // Look for chained method calls by traversing up the AST
        // We need to find the full chain starting from .styles()
        const chainedCalls: any[] = [];
        let currentPath: any = path;

        // Traverse up to find all chained calls
        while (currentPath) {
          if (
            t.isCallExpression(currentPath.node) &&
            t.isMemberExpression(currentPath.node.callee) &&
            t.isIdentifier(currentPath.node.callee.property)
          ) {
            chainedCalls.push({
              method: currentPath.node.callee.property.name,
              args: currentPath.node.arguments,
            });
          }

          // Check if parent is also a call expression (chained)
          if (
            currentPath.parentPath &&
            t.isMemberExpression(currentPath.parentPath.node) &&
            currentPath.parentPath.parentPath &&
            t.isCallExpression(currentPath.parentPath.parentPath.node)
          ) {
            currentPath = currentPath.parentPath.parentPath;
          } else {
            break;
          }
        }

        // Process chained calls in order
        const variantCalls: any[] = [];

        for (const call of chainedCalls) {
          if (call.method === 'variant' && call.args[0]) {
            const variantArg = call.args[0];
            if (t.isObjectExpression(variantArg)) {
              variantCalls.push(extractObjectLiteral(variantArg));
            }
          } else if (call.method === 'states' && call.args[0]) {
            const statesArg = call.args[0];
            if (t.isObjectExpression(statesArg)) {
              extracted.states = extractObjectLiteral(statesArg);
            }
          } else if (call.method === 'groups' && call.args[0]) {
            const groupsArg = call.args[0];
            if (t.isObjectExpression(groupsArg)) {
              const groups = extractObjectLiteral(groupsArg);
              extracted.groups = Object.keys(groups).filter(
                (key) => groups[key]
              );
            }
          } else if (call.method === 'props' && call.args[0]) {
            const propsArg = call.args[0];
            if (t.isObjectExpression(propsArg)) {
              extracted.props = extractObjectLiteral(propsArg);
            }
          }
        }

        // Store variants - if multiple, keep as array; if single, flatten
        if (variantCalls.length > 1) {
          extracted.variants = variantCalls;
        } else if (variantCalls.length === 1) {
          extracted.variants = variantCalls[0];
        }

        extractedComponents.push(extracted);
      }
    },
  });

  return extractedComponents;
}

/**
 * Extract a literal object from an AST ObjectExpression
 */
function extractObjectLiteral(node: t.ObjectExpression): Record<string, any> {
  const result: Record<string, any> = {};

  for (const prop of node.properties) {
    if (t.isObjectProperty(prop) && !prop.computed) {
      const key = t.isIdentifier(prop.key)
        ? prop.key.name
        : t.isStringLiteral(prop.key)
          ? prop.key.value
          : null;

      if (key) {
        const value = extractValue(prop.value);
        if (value !== undefined) {
          result[key] = value;
        }
      }
    }
  }

  return result;
}

/**
 * Extract a literal value from an AST node
 */
function extractValue(node: t.Node): any {
  if (t.isStringLiteral(node)) {
    return node.value;
  } else if (t.isNumericLiteral(node)) {
    return node.value;
  } else if (t.isBooleanLiteral(node)) {
    return node.value;
  } else if (t.isNullLiteral(node)) {
    return null;
  } else if (t.isObjectExpression(node)) {
    return extractObjectLiteral(node);
  } else if (t.isArrayExpression(node)) {
    return node.elements
      .filter((el): el is t.Expression | t.SpreadElement => el !== null)
      .map((el) => (t.isExpression(el) ? extractValue(el) : undefined))
      .filter((v) => v !== undefined);
  } else if (t.isTemplateLiteral(node) && node.expressions.length === 0) {
    // Simple template literal with no expressions
    return node.quasis[0].value.cooked;
  }

  // Return undefined for non-literal expressions
  return undefined;
}
