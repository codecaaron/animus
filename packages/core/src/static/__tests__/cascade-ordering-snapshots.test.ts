import { extractStylesFromCode } from '../extractor';
import { CSSGenerator } from '../generator';

import { ComponentRegistry } from '../component-registry';
import type { ComponentEntry } from '../component-registry';
import type { ExtractedStylesWithIdentity } from '../component-identity';

describe('CSS Cascade Ordering Snapshots', () => {
  describe('Grouped Mode Cascade', () => {
    const generator = new CSSGenerator({ atomic: false });

    it('should order styles, variants, and states correctly', () => {
      const code = `
        const Button = animus
          .styles({
            padding: 16,
            bg: 'gray',
            '&:hover': {
              bg: 'darkgray',
            }
          })
          .variant({
            prop: 'size',
            variants: {
              small: { padding: 8 },
              large: { padding: 24 },
            }
          })
          .variant({
            prop: 'color',
            variants: {
              primary: { bg: 'blue', color: 'white' },
              danger: { bg: 'red', color: 'white' },
            }
          })
          .states({
            disabled: { opacity: 0.5, cursor: 'not-allowed' },
            loading: { cursor: 'wait' },
          })
          .asElement('button');
      `;

      const extracted = extractStylesFromCode(code);
      const result = generator.generateFromExtracted(extracted[0]);

      expect(result.css).toMatchSnapshot();
    });

    it('should handle responsive values within each cascade level', () => {
      const code = `
        const Card = animus
          .styles({
            padding: { _: 16, sm: 24, lg: 32 },
            display: 'block',
          })
          .variant({
            prop: 'layout',
            variants: {
              horizontal: {
                display: 'flex',
                gap: { _: 12, md: 20 },
              }
            }
          })
          .states({
            expanded: {
              padding: { _: 24, sm: 32, lg: 48 },
            }
          })
          .asElement('div');
      `;

      const extracted = extractStylesFromCode(code);
      const result = generator.generateFromExtracted(extracted[0]);

      expect(result.css).toMatchSnapshot();
    });

    it('should maintain variant order from builder chain', () => {
      const code = `
        const Component = animus
          .styles({ color: 'black' })
          .variant({
            prop: 'tone',
            variants: {
              light: { bg: 'white' },
              dark: { bg: 'black' },
            }
          })
          .variant({
            prop: 'tone',  // Same prop, different values - should cascade
            variants: {
              light: { borderColor: 'gray' },
              dark: { borderColor: 'white' },
            }
          })
          .asElement('div');
      `;

      const extracted = extractStylesFromCode(code);
      const result = generator.generateFromExtracted(extracted[0]);

      expect(result.css).toMatchSnapshot();
    });

    it('should handle complex real-world component with all features', () => {
      const code = `
        const ArticleCard = animus
          .styles({
            // Responsive spacing
            p: { _: 16, sm: 24, lg: 32 },
            m: [0, 8, 16, 24],
            
            // Layout
            display: 'flex',
            flexDirection: { _: 'column', md: 'row' },
            gap: { _: 16, sm: 24 },
            
            // Styling
            bg: 'white',
            borderRadius: 8,
            boxShadow: '0 2px 8px rgba(0,0,0,0.1)',
            
            // Responsive sizing
            maxWidth: { _: '100%', sm: 640, lg: 960 },
            mx: 'auto',
            
            // Hover state
            '&:hover': {
              boxShadow: '0 4px 16px rgba(0,0,0,0.2)',
              transform: 'translateY(-2px)',
            }
          })
          .variant({
            prop: 'size',
            variants: {
              compact: {
                p: { _: 12, sm: 16 },
                gap: 12,
              },
              spacious: {
                p: { _: 24, sm: 32, lg: 48 },
                gap: { _: 24, md: 32 },
              }
            }
          })
          .variant({
            prop: 'theme',
            variants: {
              light: {
                bg: 'white',
                color: 'black',
                borderWidth: 1,
                borderColor: 'gray.200',
              },
              dark: {
                bg: 'gray.900',
                color: 'white',
                borderWidth: 1,
                borderColor: 'gray.700',
              }
            }
          })
          .states({
            featured: {
              borderWidth: 2,
              borderColor: 'primary',
              p: { _: 24, sm: 32, lg: 40 },
              '&::before': {
                content: '""',
                position: 'absolute',
                inset: 0,
                bg: 'primary',
                opacity: 0.1,
              }
            },
            disabled: {
              opacity: 0.5,
              cursor: 'not-allowed',
              '&:hover': {
                transform: 'none',
                boxShadow: '0 2px 8px rgba(0,0,0,0.1)',
              }
            }
          })
          .groups({
            space: true,
            color: true,
          })
          .props({
            elevation: {
              property: 'boxShadow',
              scale: {
                low: '0 1px 3px rgba(0,0,0,0.1)',
                medium: '0 4px 6px rgba(0,0,0,0.1)',
                high: '0 10px 20px rgba(0,0,0,0.1)',
              }
            }
          })
          .asElement('article');
      `;

      const extracted = extractStylesFromCode(code);
      const result = generator.generateFromExtracted(extracted[0]);

      expect(result.css).toMatchSnapshot();
    });
  });

  describe('Component Styles (Always Grouped)', () => {
    const generator = new CSSGenerator({ atomic: true }); // atomic flag is ignored for component styles

    it('should generate grouped styles even when atomic flag is set', () => {
      const code = `
        const Button = animus
          .styles({
            padding: { _: 16, sm: 24 },
            margin: 0,
            bg: 'blue',
            '&:hover': {
              bg: 'darkblue',
            }
          })
          .variant({
            prop: 'size',
            variants: {
              small: { padding: 8, fontSize: 14 },
              large: { padding: 32, fontSize: 20 },
            }
          })
          .states({
            disabled: { 
              opacity: 0.5,
              cursor: 'not-allowed',
            },
          })
          .asElement('button');
      `;

      const extracted = extractStylesFromCode(code);
      const result = generator.generateFromExtracted(extracted[0]);

      expect(result.css).toMatchSnapshot();
    });
  });
});
