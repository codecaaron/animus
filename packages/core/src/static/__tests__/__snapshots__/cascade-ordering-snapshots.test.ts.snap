// Jest Snapshot v1, https://jestjs.io/docs/snapshot-testing

exports[`CSS Cascade Ordering Snapshots Component Styles (Always Grouped) should generate grouped styles even when atomic flag is set 1`] = `
"/* Base Styles */
.animus-Button-b6n {
  padding: 16px;
  margin: 0px;
  background-color: blue;
}
.animus-Button-b6n:hover {
  background-color: darkblue;
}
@media screen and (min-width: 768px) {
  .animus-Button-b6n {
    padding: 24px;
  }
}

/* Variant: size="small" */
.animus-Button-b6n-size-small {
  padding: 8px;
  font-size: 14px;
}

/* Variant: size="large" */
.animus-Button-b6n-size-large {
  padding: 32px;
  font-size: 20px;
}

/* State: disabled */
.animus-Button-b6n-state-disabled {
  opacity: 0.5;
  cursor: not-allowed;
}"
`;

exports[`CSS Cascade Ordering Snapshots Grouped Mode Cascade should handle complex real-world component with all features 1`] = `
"/* Base Styles */
.animus-ArticleCard-a11d {
  padding: 16px;
  margin: 0px;
  gap: 16px;
  display: flex;
  flex-direction: column;
  background-color: white;
  border-radius: 8px;
  box-shadow: 0 2px 8px rgba(0,0,0,0.1);
  max-width: 100%;
  margin-left: auto;
  margin-right: auto;
}
.animus-ArticleCard-a11d:hover {
  box-shadow: 0 4px 16px rgba(0,0,0,0.2);
  transform: translateY(-2px);
}
@media screen and (min-width: 768px) {
  .animus-ArticleCard-a11d {
    padding: 24px;
    margin: 16px;
    gap: 24px;
    max-width: 640px;
  }
}
@media screen and (min-width: 1200px) {
  .animus-ArticleCard-a11d {
    padding: 32px;
    max-width: 960px;
  }
}
@media screen and (min-width: 480px) {
  .animus-ArticleCard-a11d {
    margin: 8px;
  }
}
@media screen and (min-width: 1024px) {
  .animus-ArticleCard-a11d {
    margin: 24px;
    flex-direction: row;
  }
}

/* Variant: size="compact" */
.animus-ArticleCard-a11d-size-compact {
  padding: 12px;
  gap: 12px;
}
@media screen and (min-width: 768px) {
  .animus-ArticleCard-a11d-size-compact {
    padding: 16px;
  }
}

/* Variant: size="spacious" */
.animus-ArticleCard-a11d-size-spacious {
  padding: 24px;
  gap: 24px;
}
@media screen and (min-width: 768px) {
  .animus-ArticleCard-a11d-size-spacious {
    padding: 32px;
  }
}
@media screen and (min-width: 1200px) {
  .animus-ArticleCard-a11d-size-spacious {
    padding: 48px;
  }
}
@media screen and (min-width: 1024px) {
  .animus-ArticleCard-a11d-size-spacious {
    gap: 32px;
  }
}

/* Variant: theme="light" */
.animus-ArticleCard-a11d-theme-light {
  border-width: 1px;
  border-color: gray.200;
  background-color: white;
  color: black;
}

/* Variant: theme="dark" */
.animus-ArticleCard-a11d-theme-dark {
  border-width: 1px;
  border-color: gray.700;
  background-color: gray.900;
  color: white;
}

/* State: featured */
.animus-ArticleCard-a11d-state-featured {
  border-width: 2px;
  border-color: primary;
  padding: 24px;
}
.animus-ArticleCard-a11d-state-featured::before {
  content: "";
  position: absolute;
  top: 0px;
  right: 0px;
  bottom: 0px;
  left: 0px;
  background-color: primary;
  opacity: 0.1;
}
@media screen and (min-width: 768px) {
  .animus-ArticleCard-a11d-state-featured {
    padding: 32px;
  }
}
@media screen and (min-width: 1200px) {
  .animus-ArticleCard-a11d-state-featured {
    padding: 40px;
  }
}

/* State: disabled */
.animus-ArticleCard-a11d-state-disabled {
  opacity: 0.5;
  cursor: not-allowed;
}
.animus-ArticleCard-a11d-state-disabled:hover {
  transform: none;
  box-shadow: 0 2px 8px rgba(0,0,0,0.1);
}


/* Atomic Utilities from Custom Props */"
`;

exports[`CSS Cascade Ordering Snapshots Grouped Mode Cascade should handle responsive values within each cascade level 1`] = `
"/* Base Styles */
.animus-Card-c4d {
  padding: 16px;
  display: block;
}
@media screen and (min-width: 768px) {
  .animus-Card-c4d {
    padding: 24px;
  }
}
@media screen and (min-width: 1200px) {
  .animus-Card-c4d {
    padding: 32px;
  }
}

/* Variant: layout="horizontal" */
.animus-Card-c4d-layout-horizontal {
  gap: 12px;
  display: flex;
}
@media screen and (min-width: 1024px) {
  .animus-Card-c4d-layout-horizontal {
    gap: 20px;
  }
}

/* State: expanded */
.animus-Card-c4d-state-expanded {
  padding: 24px;
}
@media screen and (min-width: 768px) {
  .animus-Card-c4d-state-expanded {
    padding: 32px;
  }
}
@media screen and (min-width: 1200px) {
  .animus-Card-c4d-state-expanded {
    padding: 48px;
  }
}"
`;

exports[`CSS Cascade Ordering Snapshots Grouped Mode Cascade should maintain variant order from builder chain 1`] = `
"/* Base Styles */
.animus-Component-c9t {
  color: black;
}

/* Variant: tone="light" */
.animus-Component-c9t-tone-light {
  background-color: white;
}

/* Variant: tone="dark" */
.animus-Component-c9t-tone-dark {
  background-color: black;
}

/* Variant: tone="light" */
.animus-Component-c9t-tone-light {
  border-color: gray;
}

/* Variant: tone="dark" */
.animus-Component-c9t-tone-dark {
  border-color: white;
}"
`;

exports[`CSS Cascade Ordering Snapshots Grouped Mode Cascade should order styles, variants, and states correctly 1`] = `
"/* Base Styles */
.animus-Button-b6n {
  padding: 16px;
  background-color: gray;
}
.animus-Button-b6n:hover {
  background-color: darkgray;
}

/* Variant: size="small" */
.animus-Button-b6n-size-small {
  padding: 8px;
}

/* Variant: size="large" */
.animus-Button-b6n-size-large {
  padding: 24px;
}

/* Variant: color="primary" */
.animus-Button-b6n-color-primary {
  background-color: blue;
  color: white;
}

/* Variant: color="danger" */
.animus-Button-b6n-color-danger {
  background-color: red;
  color: white;
}

/* State: disabled */
.animus-Button-b6n-state-disabled {
  opacity: 0.5;
  cursor: not-allowed;
}

/* State: loading */
.animus-Button-b6n-state-loading {
  cursor: wait;
}"
`;
